src/endpoint/FaabricEndpoint.cpp

```diff
@@ -1,92 +1,261 @@
 #include <faabric/endpoint/FaabricEndpoint.h>
 #include <faabric/endpoint/FaabricEndpointHandler.h>
-#include <faabric/util/config.h>
-#include <faabric/util/locks.h>
-#include <faabric/util/logging.h>
-#include <faabric/util/timing.h>
+#include <faabric/scheduler/Scheduler.h>
 
+#include <functional>
+#include <optional>
 #include <signal.h>
+#include <stdexcept>
+#include <thread>
+#include <vector>
 
 namespace faabric::endpoint {
 
-FaabricEndpoint::FaabricEndpoint()
-  : FaabricEndpoint(faabric::util::getSystemConfig().endpointPort,
-                    faabric::util::getSystemConfig().endpointNumThreads)
-{}
+namespace detail {
+struct EndpointState
+{
+    EndpointState(int threadCountIn)
+      : ioc(threadCountIn)
+    {}
+    asio::io_context ioc;
+    std::vector<std::jthread> ioThreads;
+};
+}
 
-FaabricEndpoint::FaabricEndpoint(int portIn, int threadCountIn)
-  : port(portIn)
-  , threadCount(threadCountIn)
-  , httpEndpoint(
-      Pistache::Address(Pistache::Ipv4::any(), Pistache::Port(portIn)))
-{}
+namespace {
+class HttpConnection : public std::enable_shared_from_this<HttpConnection>
+{
+    asio::io_context& ioc;
+    beast::tcp_stream stream;
+    beast::flat_buffer buffer;
+    beast::http::request_parser<beast::http::string_body> parser;
+    std::shared_ptr<HttpRequestHandler> handler;
 
-void FaabricEndpoint::start(EndpointMode mode)
+  public:
+    HttpConnection(asio::io_context& iocIn,
+                   asio::ip::tcp::socket&& socket,
+                   std::shared_ptr<HttpRequestHandler> handlerIn)
+      : ioc(iocIn)
+      , stream(std::move(socket))
+      , buffer()
+      , parser()
+      , handler(handlerIn)
+    {}
+
+    void run()
```

Comment:
I am a bit confused, why should an `HttpConnection` have a `run()` method?

Reply:
OK, I see that an `HttpConnection` is what we run on `accept`. I don't know if `run` is the best name, but maybe with a bit of explanation at the begining of the class I would have been all right.